src/python-bindings/htcondor2/credd.cpp

```diff
@@ -1 +1,175 @@
+    return true;
+}
+
+
+static PyObject *
+_credd_do_store_cred(PyObject *, PyObject * args) {
+    // _credd_do_store_cred(addr, user, credential, mode, service, handle)
+
+    const char * addr = NULL;
+    const char * user = NULL;
+    const char * credential = NULL;
+    Py_ssize_t credential_len = 0;
+    const char * service = NULL;
+    const char * handle = NULL;
+    long mode = 0;
+    if(! PyArg_ParseTuple( args, "zzz#lzz", & addr, & user, & credential, & credential_len, & mode, & service, & handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    std::string cooked_user;
+    if(! cook_user(user, mode, cooked_user)) {
+        // This was HTCondorValueError in version 1.
+        PyErr_SetString( PyExc_ValueError, "invalid user argument" );
+        return NULL;
+    }
+    // This version of do_store_cred() has been deprecated, but was used
+    // in version 1 by the [add|delete|query]_password() functions.  If
+    // the new version, below, doesn't work, that's a bug in C++ code.
+    // However, the dprint() logs will still differ from version 1.
+    //
+    // int result = do_store_cred(cooked_user.c_str(), credential, mode, & d);
+
+    ClassAd * serviceAd = NULL;
+    if( service != NULL ) {
+        serviceAd = new ClassAd();
+
+        serviceAd->Assign("service", service);
+        if( handle != NULL ) {
+            serviceAd->Assign("handle", handle);
+        }
+    } else if( handle != NULL ) {
+        // This was HTCondorValueError in version 1.
+        PyErr_SetString( PyExc_ValueError, "invalid service argument" );
+        return NULL;
+    }
+
+    Daemon * d = NULL;
+    if( addr != NULL ) {
+        d = new Daemon(DT_CREDD, addr);
+    }
+
+    ClassAd returnAd;
+    int result = do_store_cred( cooked_user.c_str(), mode,
+        (const unsigned char *)credential, credential_len,
+        returnAd, serviceAd, d);
+    if( d != NULL ) { delete d; }
+    if( serviceAd != NULL ) { delete serviceAd; }
+
+    const int MODE_MASK = 3;
+    bool query_mode = (mode & MODE_MASK) == GENERIC_QUERY;
+    bool delete_mode = (mode & MODE_MASK) == GENERIC_DELETE;
+    if( result == FAILURE_NOT_FOUND && (query_mode || delete_mode) ) {
+        return PyLong_FromLong(result);
+    }
+
+    const char * errString = NULL;
+    if( _store_cred_failed(result, mode, & errString) ) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, errString );
+        return NULL;
+    }
+
+    const int TYPE_MASK = 0x2c;
+    if( (mode & MODE_MASK) == GENERIC_QUERY &&
+      (mode & TYPE_MASK) == STORE_CRED_USER_OAUTH ) {
+        std::string str;
+        sPrintAd(str, returnAd);
+        return PyUnicode_FromString(str.c_str());
+    }
+
+    return PyLong_FromLong(result);
+}
+
+
+static PyObject *
+_credd_do_check_oauth_creds(PyObject *, PyObject * args) {
+    // _credd_do_check_oauth_creds(addr, user, serviceAds)
```

Comment:
This comment is missing the mode argument

Reply:
Fixed.